index.js

```diff
@@ -0,0 +1,247 @@
+                }
+                const options = hoek.applyToDefaults({ ecosystem },
+                    (scm.config || {}));  // Add ecosystem to scm options
+
+                this.loadPlugin(scm.plugin, options);
+            });
+        }
+
+        if (Object.keys(this.scms).length === 0) {
+            throw new Error('No scm config passed in.');
+        }
+    }
+
+    loadPlugin(plugin, options) {
+        // eslint-disable-next-line global-require, import/no-dynamic-require
+        const ScmPlugin = require(`screwdriver-scm-${plugin}`);
+        const scmPlugin = new ScmPlugin(options);
+        const scmContexts = scmPlugin.getScmContexts();
+        const scmContext = scmContexts[0]; // plugins return only one scmContext
+
+        if (scmContext == null || scmContext.length === 0) {
+            console.warn('Illegal scmContext plugin.');
+
+            return;
+        }
+
+        if (typeof this.scms[scmContext] === 'object') {
+            console.warn('Duplicate scm config.');
+        }
+
+        this.scms[scmContext] = scmPlugin;
+    }
+
+    /**
+     * Process by choosen webhook scm module
+     * @method chooseScm
+     * @param  {Object}     headers          The request headers associated with the webhook payload
+     * @param  {Object}     payload          The webhook payload received from the SCM service
+     * @param  {Function}   callback         fn(scm)
+     * @return {Promise}                     callback result
+     */
+    chooseWebhookScm(headers, payload, callback) {
+        return new Promise((resolve, reject) => {
+            // choose a webhook scm module, or undefined if there is no suitable one
+            async.detectSeries(this.scms, (scm, cb) => {
+                scm.canHandleWebhook(headers, payload)
+                .then((result) => {
+                    cb(result === false ? null : scm);
+                }).catch(() => {
+                    cb(null);
+                });
+            }, (ret) => {
+                if (ret == null) {
+                    reject('there is no suitable webhook module');
+                }
+
+                resolve(callback(ret));
+            });
+        });
+    }
+
+    /**
+     * Process by choosen scm module
+     * @method chooseScm
+     * @param  {Object}     config              Configuration
+     * @param  {String}     config.scmContext   Name of scm context
+     * @param  {Function}   callback            fn(scm)
+     * @return {Promise}                        callback result
+     */
+    chooseScm(config, callback) {
+        return new Promise((resolve, reject) => {
+            if (config.scmContext == null || config.scmContext.length === 0) {
+                reject('Not implemented');
+            }
+
+            const scm = this.scms[config.scmContext];
+
+            if (typeof scm !== 'object') {
+                reject('Not implemented');
+            }
+
+            resolve(callback(scm));
```

Comment:
You can omit an use of callback function if `chooseScm` resolves to scm object, and the promise finally resolves to the result of a function in each scm module.
```
return this.chooseScm(config).then(scm => scm.getCommitSha(config));
```